compiler/src/dotty/tools/backend/sjs/JUnitBootstrappers.scala

```diff
@@ -0,0 +1,341 @@
+
+    lazy val TestMetadataType: TypeRef = ctx.requiredClassRef("org.scalajs.junit.TestMetadata")
+
+    lazy val NoSuchMethodExceptionType: TypeRef = ctx.requiredClassRef("java.lang.NoSuchMethodException")
+
+    lazy val FutureType: TypeRef = ctx.requiredClassRef("scala.concurrent.Future")
+    def FutureClass(implicit ctx: Context): ClassSymbol = FutureType.symbol.asClass
+
+    private lazy val FutureModule_successfulR = ctx.requiredModule("scala.concurrent.Future").requiredMethodRef("successful")
+    def FutureModule_successful(implicit ctx: Context): Symbol = FutureModule_successfulR.symbol
+
+    private lazy val SuccessModule_applyR = ctx.requiredModule("scala.util.Success").requiredMethodRef(nme.apply)
+    def SuccessModule_apply(implicit ctx: Context): Symbol = SuccessModule_applyR.symbol
+  }
+
+  // The actual transform -------------------------------
+
+  override def transformPackageDef(tree: PackageDef)(implicit ctx: Context): Tree = {
+    // TODO Can we cache this better?
+    implicit val mydefn: MyDefinitions = new MyDefinitions()
+
+    @tailrec
+    def hasTests(sym: ClassSymbol): Boolean = {
+      sym.info.decls.exists(m => m.is(Method) && m.hasAnnotation(mydefn.TestAnnotClass)) ||
+      sym.superClass.exists && hasTests(sym.superClass.asClass)
+    }
+
+    def isTestClass(sym: Symbol): Boolean = {
+      sym.isClass &&
+      !sym.is(ModuleClass | Abstract | Trait) &&
+      hasTests(sym.asClass)
+    }
+
+    val bootstrappers = tree.stats.collect {
+      case clDef: TypeDef if isTestClass(clDef.symbol) =>
+        genBootstrapper(clDef.symbol.asClass)
+    }
+
+    if (bootstrappers.isEmpty) tree
+    else cpy.PackageDef(tree)(tree.pid, tree.stats ::: bootstrappers)
+  }
+
+  private def genBootstrapper(testClass: ClassSymbol)(
+      implicit ctx: Context, mydefn: MyDefinitions): TypeDef = {
+
+    val owner = testClass.owner
+    val moduleSym = ctx.newCompleteModuleSymbol(owner,
+        (testClass.name ++ "$scalajs$junit$bootstrapper").toTermName,
+        Synthetic, Synthetic,
+        List(defn.ObjectType, mydefn.BootstrapperType), newScope,
+        coord = testClass.span, assocFile = testClass.assocFile).entered
+    val classSym = moduleSym.moduleClass.asClass
+
+    val constr = genConstructor(classSym)
+
+    val testMethods = annotatedMethods(testClass, mydefn.TestAnnotClass)
+
+    val defs = List(
+        genCallOnModule(classSym, Names.beforeClass, testClass.companionModule, mydefn.BeforeClassAnnotClass),
+        genCallOnModule(classSym, Names.afterClass, testClass.companionModule, mydefn.AfterClassAnnotClass),
+        genCallOnParam(classSym, Names.before, testClass, mydefn.BeforeAnnotClass),
+        genCallOnParam(classSym, Names.after, testClass, mydefn.AfterAnnotClass),
+        genTests(classSym, testMethods),
+        genInvokeTest(classSym, testClass, testMethods),
+        genNewInstance(classSym, testClass)
+    )
+
+    if (ctx.sbtCallback != null)
```

Comment:
I wonder if we shouldn't move that check inside `registerDummyClass` ? Otherwise every caller will need to duplicate it.